### PR TITLE
[CP] Send email to notify adherent citizen project creation (approve)

### PR DESCRIPTION
**Pull request by NyKo24**

### CODE REVIEW

File: src/Consumer/ProjectCitizenCreationNotificationConsumer.php [hunk 1]
```diff
@@ -57,32 +55,36 @@ protected function doExecute(array $data): int
...
 
             return self::MSG_ACK;
         }
 
         $offset = $data['offset'];
         $totalAdherent = $this->citizenProjectManager->findAdherentNearCitizenProjectOrAcceptAllNotification($citizenProject, $offset)->count();
 
-        $this->writeln('info', sprintf('Start sending. offset : %s | totalAdherent : %s | citizenProjectUuid', $offset, $totalAdherent, $citizenProject->getUuid()->toString()));
+        if (!$totalAdherent) {
+            $this->writeln('info', 'No adherent to notify foud for '.$data['uuid'].' citizen project');
```
Comment (ottaviano) [inline src/Consumer/ProjectCitizenCreationNotificationConsumer.php]: found

File: src/Form/AdherentEmailSubscriptionType.php [hunk 2]
```diff
@@ -2,27 +2,50 @@
...
+                        'choices' => AdherentEmailSubscription::SUBSCRIPTIONS,
+                        'strict' => true,
+                    ])],
+                ]),
+                'expanded' => true,
+                'multiple' => true,
+            ])
+            ->add('citizenProjectCreationEmailSubscriptionRadius', ChoiceType::class, [
+                'choices' => AdherentEmailSubscription::CITIZEN_PROJECT_DISTANCE_NOTIFICATION,
+                'label' => 'Préference notification',
```
Comment (ottaviano) [inline src/Form/AdherentEmailSubscriptionType.php]: Préférence

File: src/Mailer/Message/CitizenProjectApprovalConfirmationMessage.php [hunk 3]
```diff
@@ -3,21 +3,23 @@
...
 
 final class CitizenProjectApprovalConfirmationMessage extends Message
 {
-    public static function create(Adherent $administrator, string $cityName, string $url): self
+    public static function create(Adherent $administrator, CitizenProject $citizenProject): self
     {
         $message = new self(
             Uuid::uuid4(),
-            '244444',
+            '244444 ',
```
Comment (ottaviano) [inline src/Mailer/Message/CitizenProjectApprovalConfirmationMessage.php]: why this change?

File: tests/CitizenProject/CitizenProjectMessageNotifierTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+use AppBundle\Entity\Adherent;
+use AppBundle\Entity\CitizenProject;
+use AppBundle\Mailer\MailerService;
+use OldSound\RabbitMqBundle\RabbitMq\ProducerInterface;
+use PHPUnit\Framework\TestCase;
+use Ramsey\Uuid\Uuid;
+use Tests\AppBundle\Controller\ControllerTestTrait;
+
+/**
+ * @functional
```
Comment (Remg) [inline tests/CitizenProject/CitizenProjectMessageNotifierTest.php]: `@group functional` ?
not needed though as these tests are unit tests

File: tests/CitizenProject/CitizenProjectMessageNotifierTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+ */
+class CitizenProjectMessageNotifierTest extends TestCase
+{
+    use ControllerTestTrait;
```
Comment (Remg) [inline tests/CitizenProject/CitizenProjectMessageNotifierTest.php]: not needed for unit tests

File: tests/CitizenProject/CitizenProjectManagerTest.php [hunk 6]
```diff
@@ -121,6 +123,33 @@ public function testChangePrivilege()
...
+
+        $this->assertSame(3, $adherents->count());
+
+        $adherents = $this->citizenProjectManager->findAdherentNearCitizenProjectOrAcceptAllNotification($citizenProject, 0, false);
+
+        $this->assertSame(4, $adherents->count());
+
+        $adherent = $this->getAdherentRepository()->findByEmail('francis.brioul@yahoo.com');
+        $adherent->setCitizenProjectCreationEmailSubscriptionRadius(AdherentEmailSubscription::DISTANCE_100KM);
+        $this->getManagerRegistry()->getManager()->persist($adherent);
```
Comment (Remg) [inline tests/CitizenProject/CitizenProjectManagerTest.php]: This is not needed since this Adherent entity is already managed.

File: tests/CitizenProject/CitizenProjectManagerTest.php [hunk 7]
```diff
@@ -121,6 +123,33 @@ public function testChangePrivilege()
...
+
+        $adherent = $this->getAdherentRepository()->findByEmail('referent@en-marche-dev.fr');
+        $adherent->setCitizenProjectCreationEmailSubscriptionRadius(AdherentEmailSubscription::DISTANCE_100KM);
+        $this->getManagerRegistry()->getManager()->persist($adherent);
```
Comment (Remg) [inline tests/CitizenProject/CitizenProjectManagerTest.php]: This is not needed since this Adherent entity is already managed.

**Comment by Remg** [comment]
`tests/CitizenProject/CitizenProjectManagerTest.php` should be annotated with `@group functional`

File: src/Form/AdherentEmailSubscriptionType.php (hunk 2 above)
Comment (NyKo24) [reply src/Form/AdherentEmailSubscriptionType.php]: ✔️

File: src/Mailer/Message/CitizenProjectApprovalConfirmationMessage.php (hunk 3 above)
Comment (NyKo24) [reply src/Mailer/Message/CitizenProjectApprovalConfirmationMessage.php]: ✔️

File: tests/CitizenProject/CitizenProjectMessageNotifierTest.php (hunk 4 above)
Comment (NyKo24) [reply tests/CitizenProject/CitizenProjectMessageNotifierTest.php]: ✔️

File: tests/CitizenProject/CitizenProjectMessageNotifierTest.php (hunk 5 above)
Comment (NyKo24) [reply tests/CitizenProject/CitizenProjectMessageNotifierTest.php]: ✔️

File: tests/CitizenProject/CitizenProjectManagerTest.php (hunk 6 above)
Comment (NyKo24) [reply tests/CitizenProject/CitizenProjectManagerTest.php]: ✔️

File: tests/CitizenProject/CitizenProjectManagerTest.php (hunk 7 above)
Comment (NyKo24) [reply tests/CitizenProject/CitizenProjectManagerTest.php]: ✔️

File: src/Consumer/ProjectCitizenCreationNotificationConsumer.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+use AppBundle\CitizenProject\CitizenProjectManager;
+use AppBundle\CitizenProject\CitizenProjectMessageNotifier;
+use AppBundle\Entity\Adherent;
+use AppBundle\Entity\CitizenProject;
+use AppBundle\Repository\CitizenProjectRepository;
+use OldSound\RabbitMqBundle\RabbitMq\ProducerInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Constraints as Assert;
+
+class ProjectCitizenCreationNotificationConsumer extends AbstractMailerConsumer
```
Comment (Remg) [inline src/Consumer/ProjectCitizenCreationNotificationConsumer.php]: please extend the `AbstractConsumer` class

File: src/Events.php [hunk 9]
```diff
@@ -15,6 +15,7 @@
     const CITIZEN_INITIATIVE_CANCELLED = 'citizen_initiativecancelled';
     const CITIZEN_PROJECT_CREATED = 'citizen_project_created';
     const CITIZEN_PROJECT_UPDATED = 'citizen_project_updated';
+    const CITIZEN_PROJECT_APPROVE = 'citizen_project_approve';
```
Comment (thomasbisignani) [inline src/Events.php]: What about to rename this const in `CITIZEN_PROJECT_APPROVED` to keep the same events logic ?

Comment (NyKo24) [reply src/Events.php]: 👍

File: src/CitizenProject/CitizenProjectCreationCommandHandler.php [hunk 10]
```diff
@@ -39,7 +38,5 @@ public function handle(CitizenProjectCreationCommand $command): void
         $this->manager->flush();
 
         $this->dispatcher->dispatch(Events::CITIZEN_PROJECT_CREATED, new CitizenProjectWasCreatedEvent($citizenProject, $adherent));
-
-        $this->mailer->sendMessage(CitizenProjectCreationConfirmationMessage::create($adherent));
```
Comment (Remg) [inline src/CitizenProject/CitizenProjectCreationCommandHandler.php]: this service no longer needs the `app.mailer.transactional` dependency

File: src/Events.php [hunk 11]
```diff
@@ -15,6 +15,7 @@
     const CITIZEN_INITIATIVE_CANCELLED = 'citizen_initiativecancelled';
     const CITIZEN_PROJECT_CREATED = 'citizen_project_created';
     const CITIZEN_PROJECT_UPDATED = 'citizen_project_updated';
+    const CITIZEN_PROJECT_APPROVED = 'citizen_project_approve';
```
Comment (Remg) [inline src/Events.php]: `citizen_project_approved`

File: templates/adherent/set_email_notifications.html.twig [hunk 12]
```diff
@@ -16,10 +16,75 @@
...
+                } else {
+                    labelCheckBoxNotificationPc.text('');
+                    labelCheckBoxNotificationPc.append(
+                        $('<span>')
+                            .append('Être notifié(e) de la création de nouveaux projets citoyens dans un rayon de ')
+                            .append(selectDistanceNotificationPc)
+                            .append('autour de chez moi')
+                    );
+                    selectDistanceNotificationPc.unbind('change').on('change', selectHandler);
+                }
```
Comment (Remg) [inline templates/adherent/set_email_notifications.html.twig]: you could avoid code duplication here by only constructing the message within the condition.

```
if (val == '0') {
    var message = 'Être notifié(e) de toute création de projet citoyen autour de chez moi';
} else {
    var message = 'Être notifié(e) de la création de nouveaux projets citoyens dans un rayon de '+selectDistanceNotificationPc+' autour de chez moi';
}

// apply text and reapply selectHandler

File: src/Events.php (hunk 11 above)
Comment (NyKo24) [reply src/Events.php]: ✔️
